crates/exex/exex/src/backfill/stream.rs

```diff
@@ -157,33 +156,44 @@ where
 
-        // Spawn new tasks only if we are below the parallelism configured.
-        while this.tasks.len() < this.parallelism {
-            // Take the next `batch_size` blocks from the range and calculate the range bounds
-            let mut range = this.range.by_ref().take(this.batch_size);
-            let start = range.next();
-            let range_bounds = start.zip(range.last().or(start));
-
-            // Create the range from the range bounds. If it is empty, we are done.
-            let Some(range) = range_bounds.map(|(first, last)| first..=last) else {
-                debug!(target: "exex::backfill", tasks = %this.tasks.len(), range = ?this.range, "No more block batches to backfill");
-                break;
-            };
-
-            // Spawn a new task for that range
-            debug!(target: "exex::backfill", tasks = %this.tasks.len(), ?range, "Spawning new block batch backfill task");
-            let job = Box::new(BackfillJob {
-                executor: this.executor.clone(),
-                provider: this.provider.clone(),
-                prune_modes: this.prune_modes.clone(),
-                thresholds: this.thresholds.clone(),
-                range,
-                stream_parallelism: this.parallelism,
-            }) as BackfillTaskIterator<_>;
-            this.push_back(job);
+        loop {
+            // Spawn new tasks only if we are below the parallelism configured.
+            while this.tasks.len() < this.parallelism {
+                // Take the next `batch_size` blocks from the range and calculate the range bounds
+                let mut range = this.range.by_ref().take(this.batch_size);
+                let start = range.next();
+                let range_bounds = start.zip(range.last().or(start));
+
+                // Create the range from the range bounds. If it is empty, we are done.
+                let Some(range) = range_bounds.map(|(first, last)| first..=last) else {
+                    debug!(target: "exex::backfill", tasks = %this.tasks.len(), range = ?this.range, "No more block batches to backfill");
+                    break;
+                };
+
+                // Spawn a new task for that range
+                debug!(target: "exex::backfill", tasks = %this.tasks.len(), ?range, "Spawning new block batch backfill task");
+                let job = Box::new(BackfillJob {
+                    executor: this.executor.clone(),
+                    provider: this.provider.clone(),
+                    prune_modes: this.prune_modes.clone(),
+                    thresholds: this.thresholds.clone(),
+                    range,
+                    stream_parallelism: this.parallelism,
+                }) as BackfillTaskIterator<_>;
+                this.push_back(job);
+            }
+
+            let res = ready!(this.poll_next_task(cx));
+
+            if res.is_some() {
+                return Poll::Ready(res);
+            }
+
+            if this.range.peek().is_none() {
```

Comment:
just https://doc.rust-lang.org/std/ops/struct.Range.html#method.is_empty should work i think?